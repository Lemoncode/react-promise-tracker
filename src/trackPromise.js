import { Emitter } from './tinyEmmiter';

export const emitter = new Emitter();
export const promiseCounterUpdateEventId = 'promise-counter-update';

let counter = 0;

// TODO: Add unit test support
export const trackPromise = (promise, areaPromise = 'global') => {
  counter++;
  const promiseInProgress = anyPromiseInProgress();
  emitter.emit(promiseCounterUpdateEventId, promiseInProgress,getArea(areaPromise));

  promise
  .then(() => decrementPromiseCounter(),
        () =>decrementPromiseCounter()
        );

  return promise;
};

const getArea = (areaPromise) => areaPromise === '' ? areaPromise='global' : areaPromise;

const anyPromiseInProgress = () => (counter > 0);

const decrementPromiseCounter = () => {
  counter--;
  const promiseInProgress = anyPromiseInProgress();
  emitter.emit(promiseCounterUpdateEventId, promiseInProgress);
};

// TODO: Enhancement we could catch here errors and throw an Event in case there's an HTTP Error
// then the consumer of this event can be listening and decide what to to in case of error
