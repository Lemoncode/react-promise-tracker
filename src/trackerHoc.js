import React, { Component } from "react";
import { emitter, promiseCounterUpdateEventId } from "./trackPromise";

export const promiseTrackerHoc = ComponentToWrap => {
  return class promiseTrackerComponent extends Component {
    constructor(props) {
      super(props);

      this.state = { trackedPromiseInProgress: false };
    }

    componentWillMount() {
      emitter.on(promiseCounterUpdateEventId, anyPromiseInProgress => {
        this.setState({ trackedPromiseInProgress: anyPromiseInProgress });
      });
    }

    componentWillUnmount() {
      emitter.off(promiseCounterUpdateEventId);
    }

    render() {
      return (
        <ComponentToWrap
          {...this.props}
          trackedPromiseInProgress={this.state.trackedPromiseInProgress}
        />
      );
    }
  };
};
