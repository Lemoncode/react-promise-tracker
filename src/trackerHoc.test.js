import React from 'react';
import { render, screen } from '@testing-library/react';
import { promiseTrackerHoc } from './trackerHoc';
import * as trackPromiseAPI from './trackPromise';
import { defaultArea } from './constants';

const TestSpinnerComponent = (props) => (
  <pre>props: {JSON.stringify(props, null, 4)}</pre>
);

describe('trackerHoc', () => {
  describe('Initial Status', () => {
    it('should render component with trackedPromiseInProgress equals false and area equals "default-area" when render promiseTrackerHoc without props', () => {
      // Arrange

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent />);

      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false, area equals "default-area" and customProp equals "test" when feeding customProp equals "test"', () => {
      // Arrange

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent customProp="test" />);

      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false and area equals "testArea" when feeding area equals "testArea"', () => {
      // Arrange

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ area: 'testArea' }} />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 0', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(0);

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(0);

      const progress = false;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false to different area', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(0);

      const progress = false;
      const area = 'otherArea';
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals true when counter is 0 and emit event with progress equals true', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(0);

      const progress = true;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals true to different area', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(0);

      const progress = true;
      const area = 'otherArea';
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals true when counter is 1', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(1);

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals true', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(1);

      const progress = true;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals true to different area', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(1);

      const progress = true;
      const area = 'otherArea';
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals false', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(1);

      const progress = false;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals false to different area', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(1);

      const progress = false;
      const area = 'otherArea';
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(<TrackedComponent />);

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false and area equals "testArea" when feeding area equals "testArea" and delay equals 300', () => {
      // Arrange

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ area: 'testArea', delay: 300 }} />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 0 and delay equals 300', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(0);

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ delay: 300 }} />
      );

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false and delay equals 300', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(0);

      const progress = false;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ delay: 300 }} />
      );

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false to different area and delay equals 300', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(0);

      const progress = false;
      const area = 'otherArea';
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ delay: 300 }} />
      );

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals true when counter is 0 and emit event with progress equals true and delay equals 300', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest
        .fn()
        .mockReturnValueOnce(0)
        .mockReturnValueOnce(1);

      const progress = true;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));
      const setTimeoutStub = jest
        .spyOn(window, 'setTimeout')
        .mockImplementation((callback) => callback());

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ delay: 300 }} />
      );

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals true to different area and delay equals 300', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(0);

      const progress = true;
      const area = 'otherArea';
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ delay: 300 }} />
      );

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 1 and delay equals 300', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(1);

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ delay: 300 }} />
      );

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals true and delay equals 300', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(1);

      const progress = true;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ delay: 300 }} />
      );

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals true to different area and delay equals 300', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(1);

      const progress = true;
      const area = 'otherArea';
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ delay: 300 }} />
      );

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals false and delay equals 300', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(1);

      const progress = false;
      const area = defaultArea;
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ delay: 300 }} />
      );

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals false to different area and delay equals 300', () => {
      // Arrange
      trackPromiseAPI.getCounter = jest.fn().mockReturnValue(1);

      const progress = false;
      const area = 'otherArea';
      const emitterStub = jest
        .spyOn(trackPromiseAPI.emitter, 'on')
        .mockImplementation((id, callback) => callback(progress, area));

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);

      // Assert
      const { asFragment } = render(
        <TrackedComponent config={{ delay: 300 }} />
      );

      expect(trackPromiseAPI.getCounter).toHaveBeenCalled();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('Handling delay timeouts', () => {
    beforeEach(() => {
      jest.useFakeTimers();
    });

    afterAll(() => {
      jest.useRealTimers();
    });

    it('should render <h1>NO SPINNER</h2> when counter is 1 but delay is set to 300 (before timing out)', async () => {
      // Arrange
      const TestSpinnerComponent = (props) => {
        return (
          <div>
            {props.promiseInProgress ? <h1>SPINNER</h1> : <h2>NO SPINNER</h2>}
          </div>
        );
      };

      const getCounterStub = jest.spyOn(trackPromiseAPI, 'getCounter');

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);
      render(<TrackedComponent config={{ delay: 300 }} />);

      // Check very beginning (no promises going on) NO SPINNER is shown
      // TODO: this assert could be skipped (move to another test)
      expect(screen.getByText('NO SPINNER')).toBeInTheDocument();
      expect(getCounterStub).toHaveBeenCalled();

      // Assert
      // This promise will resolved after 1 seconds, by doing this
      // we will be able to test 2 scenarios:
      // [0] first 200ms spinner won't be shown (text NOSPINNER)
      // [1] after 200ms spinner will be shown (text SPINNER)
      // [2] after 1000ms spinner will be hidded again (text NOSPINNER)

      const myFakePromise = new Promise((resolve) => {
        setTimeout(() => {
          resolve(true);
        }, 1000);
      });

      trackPromiseAPI.trackPromise(myFakePromise);

      jest.advanceTimersByTime(100);

      // [0] first 200ms spinner won't be shown (text NOSPINNER)
      expect(await screen.findByText('NO SPINNER')).toBeInTheDocument();

      jest.advanceTimersByTime(300);

      // Before the promise get's resolved
      // [1] after 200ms spinner will be shown (text SPINNER)
      expect(await screen.findByText('SPINNER')).toBeInTheDocument();

      // After the promise get's resolved
      jest.runAllTimers();

      // [2] after 1000ms spinner will be hidded again (text NOSPINNER)
      // Wait for fakePromise (simulated ajax call) to be completed
      // no spinner should be shown

      await myFakePromise;
      expect(await screen.findByText('NO SPINNER')).toBeInTheDocument();
      expect(screen.queryByText('SPINNER')).not.toBeInTheDocument();
    });

    it('should render <h1>SPINNER</h2> when counter is 1, delay is set to 1000 and promise has 2000 timeout', async () => {
      // Arrange
      const TestSpinnerComponent = (props) => {
        return (
          <div>
            {props.promiseInProgress ? <h1>SPINNER</h1> : <h2>NO SPINNER</h2>}
          </div>
        );
      };

      const getCounterStub = jest.spyOn(trackPromiseAPI, 'getCounter');

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);
      render(<TrackedComponent config={{ delay: 1000 }} />);

      expect(await screen.findByText('NO SPINNER')).toBeInTheDocument();
      expect(getCounterStub).toHaveBeenCalled();

      // Assert
      const myFakePromise = new Promise((resolve) => {
        setTimeout(() => {
          resolve(true);
        }, 2000);
      });

      trackPromiseAPI.trackPromise(myFakePromise);

      jest.advanceTimersByTime(500);
      expect(await screen.findByText('NO SPINNER')).toBeInTheDocument();

      // Total advance 1000
      jest.advanceTimersByTime(500);
      expect(await screen.findByText('SPINNER')).toBeInTheDocument();

      // Total advance 1999
      jest.advanceTimersByTime(999);
      expect(await screen.findByText('SPINNER')).toBeInTheDocument();

      // After the promise get's resolved
      jest.runAllTimers();

      await myFakePromise;
      expect(await screen.findByText('NO SPINNER')).toBeInTheDocument();

      // Total advance 2010
      jest.advanceTimersByTime(11);
      expect(await screen.findByText('NO SPINNER')).toBeInTheDocument();
    });

    it('should render <h1>NO SPINNER</h2> when counter is 1, delay is set to 2000 and promise has 1000 timeout', async () => {
      // Arrange
      const TestSpinnerComponent = (props) => {
        return (
          <div>
            {props.promiseInProgress ? <h1>SPINNER</h1> : <h2>NO SPINNER</h2>}
          </div>
        );
      };

      const getCounterStub = jest.spyOn(trackPromiseAPI, 'getCounter');

      // Act
      const TrackedComponent = promiseTrackerHoc(TestSpinnerComponent);
      render(<TrackedComponent config={{ delay: 2000 }} />);

      expect(await screen.findByText('NO SPINNER')).toBeInTheDocument();
      expect(getCounterStub).toHaveBeenCalled();

      // Assert
      const myFakePromise = new Promise((resolve) => {
        setTimeout(() => {
          resolve(true);
        }, 1000);
      });

      trackPromiseAPI.trackPromise(myFakePromise);

      jest.advanceTimersByTime(500);
      expect(await screen.findByText('NO SPINNER')).toBeInTheDocument();

      // Total advance 1000
      jest.advanceTimersByTime(500);
      expect(await screen.findByText('NO SPINNER')).toBeInTheDocument();

      await myFakePromise;
      expect(await screen.findByText('NO SPINNER')).toBeInTheDocument();

      // Total advance 1999
      jest.advanceTimersByTime(999);
      expect(await screen.findByText('NO SPINNER')).toBeInTheDocument();

      // After the promise get's resolved
      jest.runAllTimers();
    });
  });
});
