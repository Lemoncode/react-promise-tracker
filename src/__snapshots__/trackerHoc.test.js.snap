// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false and area equals "default-area" when render promiseTrackerHoc without props 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false and area equals "testArea" when feeding area equals "testArea" 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "testArea",
        "delay": 0
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false and area equals "testArea" when feeding area equals "testArea" and delay equals 300 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "testArea",
        "delay": 300
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 0 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 0 and delay equals 300 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 300
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false and delay equals 300 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 300
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false to different area 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals false to different area and delay equals 300 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 300
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals true to different area 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 0 and emit event with progress equals true to different area and delay equals 300 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 300
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 1 and delay equals 300 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 300
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals false 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals false and delay equals 300 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 300
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals false to different area and delay equals 300 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 300
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals true and delay equals 300 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 300
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false when counter is 1 and emit event with progress equals true to different area and delay equals 300 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 300
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals false, area equals "default-area" and customProp equals "test" when feeding customProp equals "test" 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "customProp": "test",
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": false
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals true when counter is 0 and emit event with progress equals true 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": true
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals true when counter is 0 and emit event with progress equals true and delay equals 300 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 300
    },
    "promiseInProgress": true
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals true when counter is 1 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": true
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals false to different area 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": true
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals true 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": true
}
  </pre>
</DocumentFragment>
`;

exports[`trackerHoc Initial Status should render component with trackedPromiseInProgress equals true when counter is 1 and emit event with progress equals true to different area 1`] = `
<DocumentFragment>
  <pre>
    props: {
    "config": {
        "area": "default-area",
        "delay": 0
    },
    "promiseInProgress": true
}
  </pre>
</DocumentFragment>
`;
